drivers/sensor/iis2iclx/iis2iclx_i2c.c

```diff
@@ -10,42 +10,24 @@
 
 #define DT_DRV_COMPAT st_iis2iclx
 
-#include <string.h>
 #include <logging/log.h>
-
 #include "iis2iclx.h"
 
 #if DT_ANY_INST_ON_BUS_STATUS_OKAY(i2c)
 
 LOG_MODULE_DECLARE(IIS2ICLX, CONFIG_SENSOR_LOG_LEVEL);
 
-static int iis2iclx_i2c_read(const struct device *dev, uint8_t reg_addr,
-			       uint8_t *value, uint8_t len)
-{
-	const struct iis2iclx_config *cfg = dev->config;
-
-	return i2c_burst_read(cfg->bus, cfg->bus_cfg.i2c_slv_addr,
-			      reg_addr, value, len);
-}
-
-static int iis2iclx_i2c_write(const struct device *dev, uint8_t reg_addr,
-				uint8_t *value, uint8_t len)
-{
-	const struct iis2iclx_config *cfg = dev->config;
-
-	return i2c_burst_write(cfg->bus, cfg->bus_cfg.i2c_slv_addr,
-			       reg_addr, value, len);
-}
-
 int iis2iclx_i2c_init(const struct device *dev)
 {
 	struct iis2iclx_data *data = dev->data;
+	const struct iis2iclx_config *cfg = dev->config;
 
-	data->ctx_i2c.read_reg = (stmdev_read_ptr) iis2iclx_i2c_read;
-	data->ctx_i2c.write_reg = (stmdev_write_ptr) iis2iclx_i2c_write;
+	/* Use generic stmemsc routine for read/write I2C bus */
+	data->ctx_i2c.read_reg = (stmdev_read_ptr) stmemsc_i2c_read;
+	data->ctx_i2c.write_reg = (stmdev_write_ptr) stmemsc_i2c_write;
 
 	data->ctx = &data->ctx_i2c;
-	data->ctx->handle = (void *)dev;
+	data->ctx->handle = (void *)&cfg->stmemsc_i2c;
```

Comment:
This is not related to this PR, but I'm a bit confused about the `ctx`, `ctx_i2c`, and `ctx_spi` members in `iis2iclx_data`.

I see that the HAL functions take an `stmdev_ctx_t*` instead of a `const stmdev_ctx_t*`, so I think I understand why this is allocated in RAM (though it's a pity because it seems like it could maybe go into ROM).

However, I don't understand why `iis2iclx_data` is allocating all three `ctx`, `ctx_i2c`, and `ctx_spi`. Why not just have:

```c
struct iis2iclx_data {
	stmdev_ctx_t ctx;
};
```

Then here (and similarly in the SPI file):

```c
data->ctx.read_reg = (stmdev_read_ptr)stmemsc_i2c_read;
data->ctx.write_reg = (stmdev_read_ptr)stmemsc_i2c_write;
data->ctx.handle = &cfg->bus_union_i_suggested_earlier.stmemsc_i2c;
```

And use `&data->ctx` wherever `data->ctx` appears today?

I feel this driver may be a basis for a regmap framework someday so I'm just asking more general questions about memory optimization.

Reply:
> This is not related to this PR, but I'm a bit confused about the `ctx`, `ctx_i2c`, and `ctx_spi` members in `iis2iclx_data`.
> 
It is not related to this PR, but I feel that you are sorting up some confusion I had about multi-instance.                                                                                                     

> I see that the HAL functions take an `stmdev_ctx_t*` instead of a `const stmdev_ctx_t*`, so I think I understand why this is allocated in RAM (though it's a pity because it seems like it could maybe go into ROM).                                                                                                     
                                                                                                     
In this case I prefer to keep this part in data, and let the fill be done in the init function (either I2C or SPI) set in configuration. This just because I'm not sure whether the interface would remain always the same and I prefer leave that part at runtime.                                                                                                     

> 
> However, I don't understand why `iis2iclx_data` is allocating all three `ctx`, `ctx_i2c`, and `ctx_spi`. Why not just have:
> 
> ```c
> struct iis2iclx_data {
> 	stmdev_ctx_t ctx;
> };
> ```
> 
> Then here (and similarly in the SPI file):
> 
> ```c
> data->ctx.read_reg = (stmdev_read_ptr)stmemsc_i2c_read;
> data->ctx.write_reg = (stmdev_read_ptr)stmemsc_i2c_write;
> data->ctx.handle = &cfg->bus_union_i_suggested_earlier.stmemsc_i2c;
> ```
> 
> And use `&data->ctx` wherever `data->ctx` appears today?
>                                                                                                      
Yes, of course. The driver structure is done in such a way we have a unique couple of (config, data) structures for each instance and stmdev_ctx_t is general, so no need for allocating separate structures for I2C and SPI. I think that initially (long back ago) the stmdev_ctx_t structure was meant to be different between the busses, and by lazyness I never went through this implementation again since that. Embarassing...

> I feel this driver may be a basis for a regmap framework someday so I'm just asking more general questions about memory optimization.

Sure. It will be the basis for a general re-structuring of stmemsc based sensor at least. Thanks for spending time on that.

Reply:
I feel I should incorporate all the suggestions in this PR

Reply:
>  It will be the basis for a general re-structuring of stmemsc based sensor at least.

The journey of a thousand miles begins with a single step. Thanks for a very nice cleanup.